Improve `SweepGradient` angle and `TileMode` documentation

SalehTZ opened the pull request:
# Improve SweepGradient and TileMode Documentation

## Description
This PR enhances the documentation for `SweepGradient` and `TileMode` to provide clearer guidance on how angles are measured and how tile modes affect sweep gradient rendering.

### Changes
1. **SweepGradient Documentation**:
   - Clarified angle measurement in radians from the positive x-axis
   - Documented angle normalization behavior for values outside [0, 2π]
   - Added detailed explanations of how each `TileMode` affects rendering outside the angular sector

2. **Gradient.sweep Constructor**:
   - Improved parameter documentation
   - Added a practical example showing how to create a 90-degree sweep gradient
   - Clarified the relationship between color stops and angles

3. **TileMode Documentation**:
   - Added sweep gradient-specific behavior to each `TileMode` variant
   - Clarified how each mode (clamp, repeated, mirror, decal) affects rendering outside the angular sector
   - Improved overall documentation structure for gradient edge behavior

## Related Issues
Fixes #166206

## Testing
- Verified documentation changes by reviewing the generated API docs
- Ensured all examples compile and render as expected

## Breaking Changes
None - this is purely a documentation improvement.

## Additional Notes
The changes make it much clearer how `startAngle` and `endAngle` interact with different `TileMode` values, which was a source of confusion in the original issue.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.


<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

SalehTZ committed:
Improve SweepGradient angle and tileMode documentation

- Clarify angle measurement in radians from positive x-axis
- Document angle normalization behavior for values outside [0, 2π]
- Explain how tileMode affects rendering outside the angular sector
- Add detailed descriptions of each tileMode's effect on sweep gradients
- Fixes #166206

SalehTZ committed:
Improve SweepGradient angle and tileMode documentation

- Clarify angle measurement in radians from positive x-axis
- Document angle normalization behavior for values outside [0, 2π]
- Explain how tileMode affects rendering outside the angular sector
- Add detailed descriptions of each tileMode's effect on sweep gradients
- Fixes #166206

SalehTZ committed:
Improve TileMode documentation for sweep gradients and ImageFilter

- Restored ImageFilter documentation
- Added detailed sweep gradient behavior for each TileMode
- Clarified how tileMode affects rendering outside the angular sector
- Maintained all existing documentation for other gradient types

SalehTZ committed:
just some here and theres

SalehTZ committed:
and more here theres

SalehTZ committed:
and more changes

abhinavs1920 commented:
@SalehTZ Awesome work on this PR! This is the perfect fix.

Clarifying the coordinate system and adding that example using TileMode.decal is exactly what was needed. This will save a lot of developers from future headaches.

Looks great to me. Nice job!

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ committed:
fix: fixed faulty lines in docs

SalehTZ committed:
fix: removed incorrect example

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4650,9 +4650,9 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// drawn between two points. In the case of radial gradients, it's the disc
-/// that covers the circle centered on a particular point up to a given radius.
+/// gradient, this is the line between the start and end points. For a radial
+/// gradient, it's the circle defined by the center and radius. For a sweep
```
`this is` vs. `it's` here?

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4650,9 +4650,9 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// that covers the circle centered on a particular point up to a given radius.
+/// gradient, this is the line between the start and end points. For a radial
+/// gradient, it's the circle defined by the center and radius. For a sweep
+/// gradient, it's the angular sector between [startAngle] and [endAngle].
```
and here

SalehTZ replied:
done

SalehTZ committed:
fix: changed `it's` to `this is`

SalehTZ committed:
Merge github.com:SalehTZ/flutter into local

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4650,9 +4650,9 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// gradient, it's between the parallel lines that are orthogonal to the line
-/// drawn between two points. In the case of radial gradients, it's the disc
-/// that covers the circle centered on a particular point up to a given radius.
+/// gradient, this is the line between the start and end points. For a radial
```
What does 'start' and 'end' here refer to? It seems like an attempt to give a specific definition to an arbitrary concept, but it isn't clear enough to be specific.

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4650,9 +4650,9 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// drawn between two points. In the case of radial gradients, it's the disc
-/// that covers the circle centered on a particular point up to a given radius.
+/// gradient, this is the line between the start and end points. For a radial
+/// gradient, this is the circle defined by the center and radius. For a sweep
```
Technically a circle is its circumference, and a disc is the area bounded by the circle on a plane.

victorsanni replied:
Overall I'm not convinced by the definition change for linear/radial gradients, maybe there is a clearer description.

SalehTZ committed:
improve docs

SalehTZ committed:
Merge branch 'enhance-SweepGradient-documentation' of github.com:SalehTZ/flutter into enhance-SweepGradient-documentation

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,31 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// This enum is used to define how the gradient or image filter should treat the regions
-/// outside that defined inner area.
+/// - **Linear gradients**: The bounds are the line segment connecting two points
+///   (typically referred to as `start` and [end](cci:1://file:///home/newelectron/dev_folder/Flutter/flutter/engine/src/flutter/lib/ui/painting.dart:439:2-479:3) in the gradient API). Colors
```
```suggestion
///   (typically referred to as `start` and `end` in the gradient API). Colors
```

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,31 @@ base class Shader extends NativeFieldWrapperClass1 {
 ///
-/// This enum is used to define how the gradient or image filter should treat the regions
-/// outside that defined inner area.
+/// - **Linear gradients**: The bounds are the line segment connecting two points
```
```suggestion
/// - **Linear gradients**: The bounds is the line segment connecting two points
```

victorsanni replied:
'bounds' is plural but 'line segment' is singular, feels a little weird. Maybe there is a better alternative to using 'bounds' here

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,31 @@ base class Shader extends NativeFieldWrapperClass1 {
+///   (typically referred to as `start` and [end](cci:1://file:///home/newelectron/dev_folder/Flutter/flutter/engine/src/flutter/lib/ui/painting.dart:439:2-479:3) in the gradient API). Colors
+///   outside this line are determined by the `TileMode`.
+///
+/// - **Radial gradients**: The bounds are the disc defined by a center and radius.
```
```suggestion
/// - **Radial gradients**: The bounds is the disc defined by a center and radius.
```

victorsanni commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,31 @@ base class Shader extends NativeFieldWrapperClass1 {
+/// - **Radial gradients**: The bounds are the disc defined by a center and radius.
+///   Colors outside this disc are determined by the `TileMode`.
+///
+/// - **Sweep gradients**: The bounds are the angular sector between `startAngle`
```
```suggestion
/// - **Sweep gradients**: The bounds is the angular sector between `startAngle`
```

victorsanni commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -984,18 +984,41 @@ class SweepGradient extends Gradient {
+  /// The gradient will be painted in the sector between [startAngle] and
+  /// [endAngle]. The [tileMode] determines what happens in the remaining area:
+  ///
+  /// - [TileMode.clamp]: The edge colors are extended to fill the remaining area.
```
Nit: Inconsistency between use of '*' and '-' here and above.

SalehTZ replied:
Thanks, done.

SalehTZ committed:
updating to suggested fixes

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

SalehTZ committed:
updating to suggested fixes

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

SalehTZ committed:
updating to suggested fixes

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

SalehTZ committed:
updating to suggested fixes

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

SalehTZ committed:
changed `-` to `*`

SalehTZ committed:
Revert "changed `-` to `*`"

This reverts commit 1bc5aca83f595594491cff90dfe68d5d78afd33e.

SalehTZ committed:
changed `-` to `*`

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

victorsanni reviewed (approved):
LGTM. Thanks for the addition.

justinmc reviewed (commented):
Thank you for cleaning up these docs! I agree this is an improvement here. The main thing I want to fix with these comments is just to distinguish between "bounds" and "inner area". Otherwise my comments are nits.

SalehTZ committed:
Refine `TileMode` documentation for gradients

The documentation for `TileMode` when used with gradients has been updated to clarify that the "bounds" define an inner area, and `TileMode` controls how colors are determined for the outer area. This is particularly relevant for radial gradients where the "bounds" is a disc.

SalehTZ commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,31 @@ base class Shader extends NativeFieldWrapperClass1 {
-/// filter, read samples over a wide area to compute the output for a given pixel. Such
-/// a filter would need to combine samples from inside the image with hypothetical
-/// color values from outside the image.
+/// Gradients are defined within specific bounds, and `TileMode` controls how colors
+/// are determined for areas outside these bounds:
```
done.

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ committed:
Merge branch 'enhance-SweepGradient-documentation' of github.com:SalehTZ/flutter into enhance-SweepGradient-documentation

SalehTZ committed:
refined `Linear gradients` related doc

SalehTZ committed:
fix: used backticks instead of Square brackets

SalehTZ commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4933,17 +4954,39 @@ base class Gradient extends Shader {
+  /// and `endAngle`. The `tileMode` determines how the gradient behaves outside
+  /// this sector.
+  ///
+  /// The [tileMode] argument specifies how the gradient should handle areas
+  /// outside the angular sector defined by [startAngle] and [endAngle]:
```
yes that was fixed. 😃👍🏻

justinmc commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4646,23 +4646,32 @@ base class Shader extends NativeFieldWrapperClass1 {
+/// - **Linear gradients**: The inner area is the area between two points
+///   (typically referred to as `start` and `end` in the gradient API), or more precisely,
+///   it's the area between the parallel lines that are orthogonal to the line drawn between the two points.
+///   Colors outside this line are determined by the `TileMode`.
```
Nit: I think "this line" should either be "these lines" or "this area", right?

SalehTZ replied:
I think `this area` is better maybe?

justinmc commented on engine/src/flutter/lib/ui/painting.dart:
```diff
@@ -4933,17 +4955,39 @@ base class Gradient extends Shader {
+  /// that specifies where `colors[i]` begins in the gradient. If `colorStops` is
   /// not provided, then only two stops, at 0.0 and 1.0, are implied (and
-  /// `color` must therefore only have two entries). Stop values less than 0.0
+  /// `colors` must therefore only have two entries). Stop values less than 0.0
```
Nit: I know this isn't your fault, but is that an umatched parenthesis `)`?

Also I think it's missing an "it" if I'm reading it right: "If `colorsStops` is `colors` it must..."

It might be nice to clean that up in this PR, but not necessary.

SalehTZ replied:
> is that an umatched parenthesis `)` ?

I don't think so, `(and 'colors' must therefore only have two entries)`, Am I missing some point?
I tweaked a little so the parenthesis are on the same line.


> I think it's missing an "it" if I'm reading it right: "If colorsStops is colors it must..."

but isn't it correct?
> If `colorStops` is provided, `colorStops[i]` is a number from 0.0 to 1.0 that specifies where `colors[i]` begins in the gradient. If `colorStops` is not provided, then only two stops, at 0.0 and 1.0, are implied (and `colors` must therefore only have two entries).

justinmc reviewed (approved):
LGTM with nits, thanks!

SalehTZ committed:
nit

SalehTZ committed:
nit: better align parentheses

SalehTZ commented:
Google test failed! IDK if I push the changes, Google test logs will disappear.

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

justinmc reviewed (approved):
Renewing my LGTM 👍 . Thanks for fixing the nits.

I assume the Google test failures were a fluke since this is a docs-only change. I'll autosubmit this.

victorsanni committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

justinmc commented:
The failing test looked like a timeout so I reran it.

victorsanni committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ committed:
Merge branch 'master' into enhance-SweepGradient-documentation

SalehTZ commented:
Phew, now it's green.
